[Persistence] Optimize gov params schema  (Issue #76)

deblasis opened the pull request:
## Description

Currently WIP and paused because of https://github.com/pokt-network/pocket/issues/76#issuecomment-1200311076

This PR changes the schema for `params`, adds `flags` (to be implemented) and aims at improving the way we update the values.

It should also improve maintainability by:
- centralizing Gov schema related metadata into the protobuf by attaching metadata to it. (see gov.proto)
- using generics for get/set methods

Fixes [issue/76](https://github.com/pokt-network/pocket/issues/76)

## Type of change
Please mark the options that are relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] Other (<_REPLACE_ME_WITH_DETAILS_>)

## How Has This Been Tested?
_[Not needed for integration PR]_

_Describe the tests and that you ran to verify your changes. If applicable, provide steps to reproduce. Bonus points for images and videos or gifs._

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md)

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] If applicable, I have made corresponding changes to related or global README
- [ ] If applicable, I have added added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] I have documented `flags` and their differences with `params` as per [comment](https://github.com/pokt-network/pocket/pull/139#issuecomment-1215602689)

deblasis committed:
feat(gov.proto): attaching metatada to generated structs in *.pb.go

Attaching metadata that can be parsed later on in order to have a single
source of truth

deblasis committed:
feat(Gov Persistence): updated schema and CRUD using metadata from proto

deblasis committed:
test(gov): updated test to use the new generic method

deblasis committed:
Merge remote-tracking branch 'origin' into issue/persistence_optimize_gov_params_schema

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis commented on persistence/db.go:
```diff
@@ -108,32 +116,48 @@ func InitializeAllTables(ctx context.Context, db *pgx.Conn) error {
+
+	//TODO (@deblasis): val_type enum
+	_, err := db.Exec(ctx, fmt.Sprintf(`%s %s`, fmt.Sprintf(CreateEnumType, schema.ValTypeName), schema.ValTypeEnumTypes))
+	if err != nil {
```
As far as I know, CREATE TYPE ... doesn't support IF EXISTS semantics so I resorted to this graceful failure handling for the specific error

deblasis commented on shared/types/genesis/proto/gov.proto:
```diff
@@ -1,124 +1,232 @@
 
 // DISCUSS(drewskey): Explore a more general purpose "feature flag" like approach for this.
 message Params {
+  //@gotags: pokt:"val_type=BIGINT"
```
Here I am attaching "metadata" to the fields so that the generated structs have them in the form of `struct tags` that we can parse similarly to what JSON marshaller/unmarshaller do basically everywhere.

Olshansk replied:
So does this mean we can use `protoc-go-inject-tag  ` to remove this hack I introduced? https://github.com/pokt-network/pocket/blob/main/shared/types/genesis/validator.go#L10

If so, we should definitely fix my hacks 😅

deblasis replied:
I'll have a look but in my comment above I wasn't referring to anything specific in the codebase but in general what the standard library (I think `encoding/json`) does, ergo: reading the struct tags that declaratively tell what to do.

:)

Also about "hacks": 
- they sent us to the moon and more.
- everything starts with a "hack"
- perfection is the enemy of action
- 🚀

Olshansk replied:
Re "hacks" - Completely agreed. Added a comment in the [other issue](https://github.com/pokt-network/pocket/issues/154#issuecomment-1208694291) so we don't forget!

deblasis commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-		app_minimum_pause_blocks_owner TEXT NOT NULL,
-		app_max_paused_blocks_owner TEXT NOT NULL,
+func init() {
+	govParamMetadataMap = parseGovProto()
```
Here I am parsing the proto file since it's not going to change at runtime.

Olshansk replied:
ACK. Please add it as a comment

deblasis commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-		params.ValidatorMinimumPauseBlocks,
-		params.ValidatorMaxPauseBlocks,
-		params.ValidatorMaximumMissedBlocks,
+	val := reflect.ValueOf(params)
```
Here we use the metadata we obtained from the proto file along with reflection to construct our SQL

Olshansk replied:
Do you think it's worth a comment that it's leveraging the tags?

deblasis replied:
Let me know if this is ok, pointing the reader to `parseGovProto` for a full description

Olshansk replied:
LGTM

deblasis commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-		params.DoubleSignBurnPercentage,
+func GetParamQuery(paramName string) string {
+	//TODO (@deblasis): clarify if and how we should use `enabled` here
+	return fmt.Sprintf(`SELECT value FROM %s WHERE name=%s AND height=%d and enabled=true`, ParamsTableName, paramName, DefaultBigInt)
```
I think I need to better understand your train of thought around the `enabled` flag. Not sure if this is correct,

Olshansk replied:
Re height - I think we'll also need to do a `where height<=%d AND enabled=true order by height desc limit 1`.

CC @andrewnguyen22  for confirmation.

deblasis commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-	)
-}
+	subQuery := fmt.Sprintf(`SELECT %s`, fields)
+	//TODO (@deblasis): clarify if and how we should use `enabled` here
```
I think I need to better understand your train of thought around the enabled flag. Not sure if this is correct,

deblasis commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-	for i, s := range paramNames {
-		if s == paramName {
-			index = i
+func parseGovProto() (govParamMetadataMap map[string]govParamMetadata) {
```
here we basically extract our custom tag `val_type` that's gonna dictate he enum we use for that specific field and also the "snake case" name that's gonna represent the "keys" we previously had defined as column names in the legacy implementation.

Olshansk commented on persistence/db.go:
```diff
@@ -108,32 +116,48 @@ func InitializeAllTables(ctx context.Context, db *pgx.Conn) error {
-	_, err := db.Exec(ctx, fmt.Sprintf(`%s %s %s`, CreateTableIfNotExists, schema.ParamsTableName, schema.ParamsTableSchema))
+
+	//TODO (@deblasis): val_type enum
+	_, err := db.Exec(ctx, fmt.Sprintf(`%s %s`, fmt.Sprintf(CreateEnumType, schema.ValTypeName), schema.ValTypeEnumTypes))
```
if we're not going to use the first value, you can just inline `; err !=nil` (see above for an example).

Ditto below.

Olshansk commented on persistence/db.go:
```diff
@@ -108,32 +116,48 @@ func InitializeAllTables(ctx context.Context, db *pgx.Conn) error {
+	if err != nil {
+		return err
+	}
+	//TODO (@deblasis): flags table
```
What exactly is the TODO if you're already creating the table?

deblasis replied:
I forgot the `//TODO remove the //TODO` 🤣

Olshansk replied:
👷

Olshansk commented on persistence/gov.go:
```diff
@@ -917,29 +916,11 @@ func (p PostgresContext) SetParam(paramName string, paramValue interface{}) erro
 	return
 }
-
-func (p PostgresContext) GetStringParam(paramName string) (s string, err error) {
```
Yes!

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
+		pnt := govParamMetadataMap[k]
+		pVal := val.Elem().FieldByName(pnt.PropertyName)
+
+		subQuery += `(`
```
NIT: I'd use a proper string builder: https://stackoverflow.com/a/1766304/768439

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-		params.ValidatorMinimumPauseBlocks,
-		params.ValidatorMaxPauseBlocks,
-		params.ValidatorMaximumMissedBlocks,
+	val := reflect.ValueOf(params)
```
NIT: Please add a 'WARNING: reflections in prod' comment.

It's always a bit scary but I like this approach much more

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
+		case ValTypeString:
+			subQuery += fmt.Sprintf("'%s', %d, true, '%s', '%s'", k, DefaultBigInt, pnt.PoktValType, pVal.Interface())
+		case ValTypeSmallInt, ValTypeBigInt:
+			subQuery += fmt.Sprintf("'%s', %d, true, '%s', %d", k, DefaultBigInt, pnt.PoktValType, pVal.Interface())
```
add a default case that logs a fatal error (i.e. panic)

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
+func extractStructTag(structTag string, key string) string {
+	for len(structTag) > 0 {
+		i := strings.IndexByte(structTag, ',')
+		if i < 0 {
```
s/`if < 0 {`/`if < 0 { // not found`

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
+		if i < 0 {
+			i = len(structTag)
+		}
+		switch s := structTag[:i]; {
```
Why even have the switch here?

deblasis replied:
Right, I simplified it. It was meant to host additional properties inside the tag but sounds like premature optimization since we might never need more than the current one here

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-		params.MessagePauseServiceNodeFee,
-		params.MessageUnpauseServiceNodeFee,
-		params.MessageChangeParameterFee,
+type ParamTypes interface {
```
s/ParamTypes/ProtobufAttributeTypes/>

deblasis replied:
`ParamTypes` represents the types that the generic can accept:
```go
type ParamTypes interface {
	int | int32 | int64 | []byte | string
}
```
technically it has nothing to do with the protobuf.
Agreed that `ParamTypes` sounds weird though

Olshansk replied:
`SupportedParamTypes` or `GovParamTypes`?

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -1,442 +1,145 @@
-func NullifyParamsQuery(height int64) string {
-	return fmt.Sprintf(`UPDATE %s SET end_height=%d WHERE end_height=%d`, ParamsTableName, height, DefaultBigInt)
+type govParamMetadata struct {
+	PoktValType  string
```
We use `PoktVal` and `Property`, but should we be consistent here like `PropertyType` and `PropertyName`, for example?

Olshansk commented:
I did a quick first review (will do an in-depth one next time) but this is frikkin awesome!

I think we should continue moving forward with it and the changes are well contained enough that I don't think the conflicts with deprecating persistence will be that bad. CC @andrewnguyen22

deblasis committed:
Merge remote-tracking branch 'origin' into issue/persistence_optimize_gov_params_schema

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(protoc-go-inject-tag): updated README.md

deblasis committed:
chore(readability): improved error handling

code review

deblasis committed:
fix(gov): bugfix + code review feedback related changes

deblasis committed:
chore(gov): code review feedback

improve readability

deblasis committed:
refactor(gov): schema.SetParam => schema.SetParamQuery

deblasis commented:
Thanks @Olshansk! I have caught up with `main` and updated a few things but I will dive deeper and also test the parts I didn't test yet (Get/Set) after your next in-depth look and which should give me more clarity around how to use `enabled` and snapshotting 🙂.

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,151 @@ package schema
+	ParamsTableSchema = `(
+		name VARCHAR(64) NOT NULL,
+		height BIGINT NOT NULL,
+		enabled BOOLEAN NOT NULL,
```
Let's remove `enabled` for params and only keep it for `flags`

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,151 @@ package schema
-		params.MessageFishermanPauseServiceNodeFee,
-		params.MessageTestScoreFee,
-		params.MessageProveTestScoreFee,
+func GetParamQuery(paramName string) string {
```
This will need  to take in a `height` param and we may need to modify the entire persistence module interface to take in a height value as well.

Regarding the query:
1. Remove enabled=true
2. Use the passed in height

cc @andrewnguyen22 to confirm

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,151 @@ package schema
-		params.MessageTestScoreFee,
-		params.MessageProveTestScoreFee,
+func GetParamQuery(paramName string) string {
+	//TODO (@deblasis): Fix this
```
We will need to differentiated between `GetParamQuery` and `GetFlagQuery` because for flags we'll want to return `value, enabled` whereas for params we only want to return `value`

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,151 @@ package schema
-		hex.EncodeToString(params.DoubleSignBurnPercentageOwner),
+func SetParamQuery[T ParamTypes](paramName string, paramValue T, height int64) string {
+	//TODO (@deblasis): Fix this
+	fields := "name,height,value,enabled,type"
```
We'll nned to generalized this since `enabled` will exist for `flags` but not for `params`.

Olshansk commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,151 @@ package schema
-	)
-}
+	subQuery := fmt.Sprintf(`SELECT %s`, fields)
+	subQuery += fmt.Sprintf(` FROM %s WHERE name='%s' AND height=%d`, ParamsTableName, paramName, height)
```
For `Get`, we do `<=height` to get the latest value from the historical state.

For `set`, we want to use the height to set the value, no need to query on it.

For `nullify`, let's delete that altogether. It's legacy and we can add it back if we need to.
- For flags, we can turn them on/off
- For params, we can set the values to 0 or -1 via Set

This will keep the design simpler for now.

Olshansk reviewed (changes requested):
Left additional comments throughout the PR (some impacting the rest of the DB).

At a high-level, the summary of the architectural changes is:
- `shared/modules/persistence_module.go` needs to be modified to accept a height for each `Get` call -  _Eventually I want to simplify that too, but we can do that later_
- Remove the `nullify` query altogether - it's legacy and we can add it back later
- Params should now have an `enabled` field but `flags` should
- Flags are not used in the database yet, but we already use it for v0 and I think most of the foundation we need for it in v1 is already in this PR.

Let me know if you need any more details and we can jump on a call!

deblasis committed:
docs(Makefile): fixed comment for go_protoc-go-inject-tag

Updated also with  so it doesn't show up in our makefile help since
it is an internal command basically

deblasis committed:
feat(Gov): fixes +  refactoring + code-review feedback

- Improved readability

 - Added scaffolding for Flags

 - Updated
schemas

 - Updated method signatures to add  (should we rename
them to GetXAt ?)

 - Added / Updated tests

 - Updated PrePersistence
just to allow compilation

deblasis committed:
Merge remote-tracking branch 'origin' into issue/persistence_optimize_gov_params_schema

deblasis committed:
feat(gov): flags implementation

flags implementation

deblasis committed:
style(gov): removed debug statement

deblasis commented on Makefile:
```diff
@@ -70,6 +70,15 @@ go_doc:
 	}
 
+.PHONY: go_protoc-go-inject-tag
+## Run `go staticcheck` on all files in the current project
```
I would say that this is more of an internal command, I updated the comment so that it is not picked up by the "help"

Olshansk replied:
Makes sense to me

deblasis commented on utility/gov.go:
```diff
@@ -10,1798 +12,475 @@ import (
 }
 
 func (u *UtilityContext) GetParamOwner(paramName string) ([]byte, error) {
+	//DISCUSS (@deblasis): here we could potentially leverage the struct tags in gov.proto by specifying an `owner` key
```
This could perhaps be optimized further by adding the appropriate attributes in our new `pokt` struct tag in `gov.proto`.

deblasis commented on utility/gov.go:
```diff
@@ -10,1798 +12,475 @@ import (
 }
 
 func (u *UtilityContext) GetParamOwner(paramName string) ([]byte, error) {
+	//DISCUSS (@deblasis): here we could potentially leverage the struct tags in gov.proto by specifying an `owner` key
+	// eg: `app_minimum_stake` could have `pokt:"owner=app_minimum_stake_owner"`
+	// in here we would use that map to point to the owner, removing this switch, centralizing the logic and making it declarative
```
This could perhaps be optimized further by adding the appropriate attributes in our new `pokt` struct tag in `gov.proto`.

Olshansk replied:
This is an interesting idea. Let's leave it as a `DISCUSS` in the code for now but ccing @andrewnguyen22 for visibility.

NIT: space before DISCUSS.

deblasis committed:
refactor(gov): refactored using stringbuilder

Olshansk commented on persistence/gov.go:
```diff
@@ -9,893 +14,89 @@ import (
-	ctx, conn, err := p.GetCtxAndConnection()
-	if err != nil {
-		return err
+func (p PostgresContext) SetFlag(paramName string, value interface{}, enabled bool) error {
```
NIT: Golang 1.18 introduce `any` so that may actually be more appropriate than `interface{}` here based on the answer in the comment here: https://stackoverflow.com/questions/72283932/go-difference-between-any-and-interface-in-varargs

deblasis replied:
I agree, I like it more as `any` here too.

I would suggest raising a separate PR in which we replace `interface{}` with `any` across the board for consistency.

Ultimately:
```go
// any is an alias for interface{} and is equivalent to interface{} in all ways.
type any = interface{}
```
but perhaps it's more idiomatic especially for newcomers to Go

Olshansk replied:
Done: https://github.com/pokt-network/pocket/issues/156

Olshansk commented on persistence/gov.go:
```diff
@@ -908,38 +109,85 @@ func (p PostgresContext) SetParam(paramName string, paramValue interface{}) erro
 }
 
-func (p PostgresContext) GetBytesParam(paramName string) (param []byte, err error) {
+func GetFlag[T int | string | []byte](p PostgresContext, paramName string, height int64) (i T, enabled bool, err error) {
```
I definitely think that the entrypoints to setting / getting the Flag / Param should definitely be different for code readability purposes, but since most of the code is the same, I think you can create a helper that takes in a function for the query creation to reduce the code footprint.

See here for an example of how we did it elsewhere: https://github.com/pokt-network/pocket/blob/main/persistence/account.go#L143

deblasis replied:
I had a look and even if I like the approach, I feel like in here it can be simplified while still keeping it DRY.

Olshansk replied:
The final result LGTM

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -28,11 +29,181 @@ func TestGetSetParam(t *testing.T) {
+	height, err := db.GetHeight()
+	require.NoError(t, err)
+
+	//maxChains, err := db.GetMaxAppChains(height)
```
remove

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -28,11 +29,181 @@ func TestGetSetParam(t *testing.T) {
+	height, err := db.GetHeight()
+	require.NoError(t, err)
+
+	//serviceNodeMinimumStake, err := db.GetParamServiceNodeMinimumStake(height)
```
rm

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -28,11 +29,181 @@ func TestGetSetParam(t *testing.T) {
+	err := db.InitParams()
+	require.NoError(t, err)
+
+	newOwner, _ := hex.DecodeString("da034209758b78eaea06dd99c07909ab54c99b44")
```
missed opportunity to have some fun :)

deblasis replied:
alright then :)

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -28,11 +29,181 @@ func TestGetSetParam(t *testing.T) {
+	height, err := db.GetHeight()
+	require.NoError(t, err)
+
+	//owner, err := db.GetServiceNodeUnstakingBlocksOwner(height)
```
rm

Olshansk replied:
Feel free to resolve the comment if there's no followup or expectation for a response.

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -28,11 +29,181 @@ func TestGetSetParam(t *testing.T) {
+	err := db.InitFlags()
+	require.NoError(t, err)
+
+	newMaxChains := 42
```
Makes sense

Olshansk reviewed (commented):
Looking solid. Code is much easier to read and understand: 
<img width="176" alt="Screen Shot 2022-08-06 at 2 39 32 PM" src="https://user-images.githubusercontent.com/1892194/183261935-6299a8c1-96e5-4dad-85a2-138057207704.png">


Left a couple of minor comments and some non-actionables. CCing @andrewnguyen22 to PTAL as well but I think we should merge this afterward.

deblasis committed:
refactor(gov): codebase footprint reduction + code-review fdbk

refactored for DRYness

 layed a couple of Easter eggs :)

 fixed
implementation details leakage

 updated tests

 nits

deblasis committed:
style(gov): scrambled 🥚🥚

hint

Olshansk commented on persistence/gov.go:
```diff
@@ -9,893 +14,85 @@ import (
+func (p PostgresContext) setParamOrFlag(name string, value any, enabled *bool) error {
+	switch t := value.(type) {
+	case int:
+		return setParamOrFlag(p, name, t, enabled)
```
Could we just use ` fallthrough` here through to the last non-default type?

deblasis replied:
This switch statement hurts my eyes too. The best way would have been if we were allowed to use a single case where `t` ended up having the matched type but that's not possible at the moment, Go reverts `t` to type `any` inside the case in that case (pun intended).

I guess Go is forcing the developer to be explicit at the cost of being... verbose.

I played a bit with `fallthrough` but I couldn't use it because of the error `MisplacedFallthrough` and after a quick search I ended up here:

https://github.com/golang/go/issues/40462

It sounds like it's still under discussion.

Olshansk commented on persistence/gov.go:
```diff
@@ -9,893 +14,85 @@ import (
 	}
-	_, err = conn.Exec(ctx, schema.InsertParams(genesis.DefaultParams()))
-	return err
+	return fmt.Errorf("how did you get here?")
```
I'll propose the best of both words: Add a break to line 87 and then do

```
return fmt.Errorf("unhandled paramType %T for value %v", t, value)
```

deblasis replied:
Yes! Me likey! We only have use `value` again instead of `t` which is undeclared outside the switch

Olshansk commented on persistence/gov.go:
```diff
@@ -9,893 +14,85 @@ import (
-// IMPROVE(team): Switch to generics
-func (p PostgresContext) SetParam(paramName string, paramValue interface{}) error {
+// setParamOrFlag sets a param or a flag.
+// If `enabled` is nil, we are dealing with a param, otherwise it's a flag
```
**NOT ACTIONABLE**: I thought about this for a moment but ultimately really like the use of the `nil` 💯

Olshansk commented on persistence/test/gov_test.go:
```diff
@@ -31,8 +32,175 @@ func TestGetSetParam(t *testing.T) {
+	err := db.InitParams()
+	require.NoError(t, err)
+
+	newOwner, _ := hex.DecodeString("=cXYUx2RvJnWjV1duBHahBlcwNFU4ZWathTU2czc0QnMYxWYwgESSxWdNJkLw4WS5AjealmWIJGSSdUY3hmVZpnWVJmaCRlTYBHbhdFZrdlbOdlYpJESjllRyM2RxITW3VleWFGctZFSwFjW2J1aW1EbUJlVKl2Tp9GSilXTHFGNxc1VzpUMSJlS5VmL5o0QWhFcrlkNJN0Y1IlbJNXSp5UMJpXVJpUaPl2YHJGaKlXZ")
```
I tried to decode it but got an error. Not sure if I'm missing something.

Might be worth doing a `require.NoError(t, err)` on the response because otherwise `newOwner` is an empty slice.


<img width="499" alt="Screen Shot 2022-08-08 at 7 12 24 PM" src="https://user-images.githubusercontent.com/1892194/183530057-f0449c07-1639-4ed7-99d8-4b70df5652f9.png">

deblasis replied:
Ops! Genuine mistake 🤦‍♂️ but good that it allowed us to spot the missing `require.NoError(t, err)`

Olshansk reviewed (approved):
**tl;dr Almost :shipit:**. Please hold off on squashing and merging

I've re-reviewed this whole PR and just left a couple of minor final NITs.

Overall, I really like the change and it'll make it much easier to both add and extend on our param+flag logic.

Going to approve this change but please hold off on @andrewnguyen22's review before submitting. After he take a look, we'll also need to make a call on whether we want to submit this before or after https://github.com/pokt-network/pocket/pull/140 as there will definitely be conflicts, but I'm sure we can figure it out.

Olshansk reviewed (changes requested):
Unapprove to avoid accidental merge.

deblasis committed:
test(gov): fix hex string :)

deblasis committed:
fix(gov): improved error handling

deblasis commented:
> Unapprove to avoid accidental merge.

This gave me the inspiration needed for this https://github.com/pokt-network/pocket/pull/160 and consequently https://github.com/pokt-network/pocket/pull/161

andrewnguyen22 commented on persistence/gov.go:
```diff
@@ -908,38 +105,51 @@ func (p PostgresContext) SetParam(paramName string, paramValue interface{}) erro
+	}
+	switch tp := any(i).(type) {
+	case int, int32, int64:
+		iConv, errr := strconv.Atoi(stringVal)
```
Why not just use `err` here?

deblasis replied:
replaced also below in line 143 `e` => `err`

andrewnguyen22 commented on utility/gov.go:
```diff
@@ -10,1798 +12,475 @@ import (
+	if ok {
+		return types.ErrUpdateParam(store.SetParam(types.BlocksPerSessionParamName, s.Value))
 	}
+	b, ok := value.(*wrapperspb.BytesValue)
```
NIT: 

If/else here, but switch statement in other places. of code:

Would prefer to keep it consistent

deblasis replied:
Nice catch, I agree!

Also, you made me look at this code long enough to realize that I was always returning `types.ErrUpdateParam(store.SetParam(types.BlocksPerSessionParamName, t.Value))` and even if `store.SetParam(types.BlocksPerSessionParamName, t.Value)` returned nil, `UpdateParam` would have returned an error regardless. Currently not caught in tests since they are failing due to planned obsolescence (related to pre_persistence).

andrewnguyen22 commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,210 @@ package schema
 	"github.com/pokt-network/pocket/shared/types/genesis"
 )
 
 // TODO(https://github.com/pokt-network/pocket/issues/76): Optimize gov parameters implementation & schema.
```
Please remove

andrewnguyen22 commented on persistence/schema/gov.go:
```diff
@@ -3,440 +3,210 @@ package schema
+//
+// It leverages metadata in the form of struct tags (see `parseGovProto` for more information).
+//
+// WARNING: reflections in prod
```
**Optional Suggestion** We could change the structure of the genesis.Params to a `[]Param` then we could loop over the param structure in order to remove reflection from prod.

I think the current approach is fine right now

andrewnguyen22 commented on shared/modules/persistence_module.go:
```diff
@@ -143,236 +143,16 @@ type PersistenceContext interface {
-	GetServiceNodeMaxPausedBlocks() (int, error)
-	GetServiceNodesPerSession() (int, error)
+	// Flags
+	InitFlags() error
```
Having a hard time finding documentation on 'flags' and why they are necessary; please add

Olshansk replied:
In the issue @deblasis had originally taken on, he actually asked about the difference (from a technical and product standpoint) between flags and params here: https://github.com/pokt-network/pocket/issues/76

I was the one who suggested that we implement the two in similar fashions, resuing a lot of code/infrastrucutre with one key difference: params are protocol level params that the drive the state of the protocol (i.e. a value such as a fee), while a flag is VERY similar but has an `enabled` flag and can, depending on the situation, determine if the value of a param is used or not but keeping the two separate.

_Perhaps we can change a flag to be a float in the future like in other feature flag systems but I think that's overkill for now_

I think we should keep it as is for now, but perhaps add a bit of documentation in a README or Changelog.

andrewnguyen22 replied:
Yeah - understood. 

Also, it might be a bit premature to add upgrade functionality for M1.

In any case, **not a blocker**. I just think it's important to have some documentation

deblasis committed:
fix(gov): fixed utility  error handling

deblasis committed:
refactor(gov): code-review feedback + nits

deblasis committed:
Merge remote-tracking branch 'origin' into issue/persistence_optimize_gov_params_schema

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(makefile): updated to reflect README changes

Olshansk commented:
@andrewnguyen22 In terms of ordering / priority, what are your thoughts on 

1. https://github.com/pokt-network/pocket/pull/140
2. This PR
3. https://github.com/pokt-network/pocket/pull/165
4. https://github.com/pokt-network/pocket/pull/166

That way we disperse the burden of merge conflicts amongst every

andrewnguyen22 commented:
I think that makes sense.

But let's talk offline per #128

Olshansk commented:
@deblasis We're likely going to merge https://github.com/pokt-network/pocket/pull/140 soon, so just a heads up that it'll require a fair amount of conflict resolution.

Otherwise, the only blocker here after synching with @andrewnguyen22 is adding some documentation explaining the different between `flags` and `params`. There is an `FAQ` section under `persistence/README.md`, so I think that'd be a good spot to add it.

deblasis commented:
> @deblasis We're likely going to merge #140 soon, so just a heads up that it'll require a fair amount of conflict resolution.

ACK. No problem, I subscribed to that issue so I have been quietly stalking you guys and I saw this coming soon my way 🙂

> Otherwise, the only blocker here after synching with @andrewnguyen22 is adding some documentation explaining the different between `flags` and `params`. There is an `FAQ` section under `persistence/README.md`, so I think that'd be a good spot to add it.

Cool, will do, thanks @Olshansk ! 
(updated PR checklist)

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/persistence_optimize_gov_params_schema

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(gov): param naming

deblasis committed:
docs(gov): Param and Flags description in README Q&A

deblasis commented on persistence/README.md:
```diff
@@ -145,6 +145,12 @@ For example, on macOS, you can check for this with `lsof -i:5432` and kill the a
 **Q**: Why not use an ORM?
 **A**: We are trying to keep the module small and lean initially but are ope
```
I noticed that the sentence here is truncated :)

Olshansk replied:
Hahaha, I think `open to suggestions` is what we wanted to say 😅

deblasis commented on persistence/README.md:
```diff
@@ -145,6 +145,12 @@ For example, on macOS, you can check for this with `lsof -i:5432` and kill the a
 **Q**: Why not use an ORM?
 **A**: We are trying to keep the module small and lean initially but are ope
 
+**Q**: What is a `Param` in the `Gov` schema?
+**A**: It represents a value associated with a name and a height that we can reference to represent governance settings. These settings have the power of altering the behaviour of various aspects of the network.
+
+**Q**: What is a `Flag` in the `Gov` schema?
+**A**: A flag is very much alike a `Param` with the difference that it also has a boolean flag to specify if that setting is enabled or not at any point in time (height). We are discussing if we should replace the boolean flag and allow multivariate feature flags.
+
```
@andrewnguyen22: I have added some color here about flags/params.

Everything (very little 😅) I know about flags comes from this comment https://github.com/pokt-network/pocket/issues/76#issue-1213446455 and subsequent conversations with @Olshansk in the form of comments to this PR. 

One thing I assumed initially, that I later changed to the current code, was that flags shouldn't have a value like params but just a boolean flag. Maybe the truth is a bit of both: we should remove the `enabled` in flags and create logic around the `value` that we can define as float to handle multivariate flags later on, perhaps this would be out of scope for this PR but I am happy to change things if needed.

As things stand, these descriptions should fit, even though we are not yet using flags anywhere so I am assuming that this is very much subject to change as we iterate regardless.

Olshansk replied:
Thanks for the extra context. I think the flag concept will definitely evolve but the current description and implementation is more than sufficient to move forward.

Olshansk reviewed (approved):
LGTM! Let's :shipit: 

I removed the `do not merge`, retriggered the jobs and am approving. 

@deblasis What we've been doing is doing a `squash-and-merge` and copy-pasting the github description into the commit.

I looked at the two issues in the jobs, and they are related to our docker images and a test that's flaky in our consensus algorithm (related to the pacemaker). Will sync with @okdas on this but it doesn't block this PR.

andrewnguyen22 commented:
Why is the automation failing on this?

Olshansk commented:
Same reasons as last night:

1. Flaky consensus test related to the pacemaker where I use timeouts. AI on @Olshansk to fix it
2. Bad credentials to log into dockerhub to build the image. AI on @okdas.

Attaching screenshots from the link above for context. I don't think either are blockers.

<img width="406" alt="Screen Shot 2022-08-17 at 9 17 30 AM" src="https://user-images.githubusercontent.com/1892194/185143949-dc65e43e-9092-462f-a91a-7045966142c8.png">
<img width="475" alt="Screen Shot 2022-08-17 at 9 17 27 AM" src="https://user-images.githubusercontent.com/1892194/185143954-078ad424-546b-465d-8031-550f82a237a9.png">
<img width="946" alt="Screen Shot 2022-08-17 at 9 17 22 AM" src="https://user-images.githubusercontent.com/1892194/185143955-1c34eda7-0e93-4a00-a24a-06d6ac7ddbae.png">
